rust/catalyst-voting/src/crypto/zk_unit_vector/decoding.rs

```diff
@@ -1,188 +1,209 @@
-        let z = Scalar::from_bytes(bytes[0..32].try_into().unwrap())
-            .map_err(|_| anyhow!("Cannot decode `z` scalar field."))?;
-        let w = Scalar::from_bytes(bytes[32..64].try_into().unwrap())
-            .map_err(|_| anyhow!("Cannot decode `w` scalar field."))?;
-        let v = Scalar::from_bytes(bytes[64..96].try_into().unwrap())
-            .map_err(|_| anyhow!("Cannot decode `v` scalar field."))?;
+impl Decode<'_, ()> for ResponseRandomness {
+    fn decode(
+        d: &mut Decoder<'_>,
+        ctx: &mut (),
+    ) -> Result<Self, minicbor::decode::Error> {
+        let z = Scalar::decode(d, ctx)?;
+        let w = Scalar::decode(d, ctx)?;
+        let v = Scalar::decode(d, ctx)?;
         Ok(Self { z, w, v })
     }
+}
+
+impl Encode<()> for UnitVectorProof {
+    fn encode<W: Write>(
+        &self,
+        e: &mut Encoder<W>,
+        ctx: &mut (),
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        if self.0.len() != self.1.len() || self.0.len() != self.2.len() {
+            return Err(minicbor::encode::Error::message(format!(
+                "All UnitVectorProof parts must have the same length: announcements = {}, choice = {}, responses = {}",
+                self.0.len(),
+                self.1.len(),
+                self.2.len()
+            )));
+        }
+
+        e.array(2)?;
+        0.encode(e, ctx)?;
+
+        e.array(self.0.len() as u64 * ITEM_ELEMENTS_LEN + 1)?;
+        for ((announcement, choice), response) in
+            self.0.iter().zip(self.1.iter()).zip(self.2.iter())
+        {
+            announcement.encode(e, ctx)?;
+            choice.first().encode(e, ctx)?;
+            choice.second().encode(e, ctx)?;
+            response.encode(e, ctx)?;
+        }
+        self.3.encode(e, ctx)
+    }
+}
 
-    /// Encode `ResponseRandomness` tos bytes.
-    #[must_use]
-    pub fn to_bytes(&self) -> [u8; Self::BYTES_SIZE] {
-        let mut res = [0; 96];
-        res[0..32].copy_from_slice(&self.z.to_bytes());
-        res[32..64].copy_from_slice(&self.w.to_bytes());
-        res[64..96].copy_from_slice(&self.v.to_bytes());
-        res
+impl Decode<'_, ()> for UnitVectorProof {
+    fn decode(
+        d: &mut Decoder<'_>,
+        ctx: &mut (),
+    ) -> Result<Self, minicbor::decode::Error> {
+        let len = decode_array_len(d, "UnitVectorProof")?;
+        if len != 2 {
+            return Err(minicbor::decode::Error::message(format!(
+                "Unexpected UnitVectorProof array length {len}, expected 2"
+            )));
+        }
+
+        let version = u64::decode(d, ctx)?;
+        if version != 0 {
+            return Err(minicbor::decode::Error::message(format!(
+                "Unexpected UnitVectorProof version value: {version}, expected 0"
+            )));
+        }
+
+        let len = decode_array_len(d, "UnitVectorProof inner array")?;
+        if len < MIN_PROOF_CBOR_ARRAY_LEN || !len.is_multiple_of(MIN_PROOF_CBOR_ARRAY_LEN) {
```

Comment:
Should not it bee something like this ?
```suggestion
        if len < MIN_PROOF_CBOR_ARRAY_LEN || !(len - 1).is_multiple_of(ITEM_ELEMENTS_LEN) {
```